File: js/lib/fpf.js
// FPF 1.1 reference implementation — encode/decode of the transport payload.
// Works unchanged in browsers (>= 2023) and Node >= 22.

// 1.1 is the only version. 1.0 was published briefly and withdrawn before any
// document was ever handed out, so it is refused rather than read.
const VERSIONS = ['1.1'];

const PREFIX_RAW = '1.';
const PREFIX_DEFLATE = '2.';

function toBase64Url(bytes) {
  let bin = '';
  for (const b of bytes) bin += String.fromCharCode(b);
  return btoa(bin).replaceAll('+', '-').replaceAll('/', '_').replace(/=+$/, '');
}

function fromBase64Url(str) {
  const b64 = str.replaceAll('-', '+').replaceAll('_', '/');
  const padded = b64 + '='.repeat((4 - (b64.length % 4)) % 4);
  const bin = atob(padded);
  return Uint8Array.from(bin, (c) => c.charCodeAt(0));
}

async function pipeThrough(bytes, transform) {
  const stream = new Blob([bytes]).stream().pipeThrough(transform);
  return new Uint8Array(await new Response(stream).arrayBuffer());
}

export async function encode(doc, { compress = true } = {}) {
  const bytes = new TextEncoder().encode(JSON.stringify(doc));
  if (!compress) return PREFIX_RAW + toBase64Url(bytes);
  const deflated = await pipeThrough(bytes, new CompressionStream('deflate-raw'));
  return PREFIX_DEFLATE + toBase64Url(deflated);
}

export async function decode(payload) {
  let bytes;
  if (payload.startsWith(PREFIX_DEFLATE)) {
    bytes = await pipeThrough(
      fromBase64Url(payload.slice(PREFIX_DEFLATE.length)),
      new DecompressionStream('deflate-raw'),
    );
  } else if (payload.startsWith(PREFIX_RAW)) {
    bytes = fromBase64Url(payload.slice(PREFIX_RAW.length));
  } else {
    throw new Error('FPF: unknown payload prefix');
  }
  return JSON.parse(new TextDecoder().decode(bytes));
}

// legal.ids carries the buyer's registration identifiers (EN 16931 BT-47), each
// qualified by an ICD scheme code (BT-47-1) drawn from the same registry as
// einvoice.eas. What a scheme means — 0002 is a French SIREN, 0009 a SIRET — and
// the check digits it implies belong to the country profiles, not here: the core
// format has no business knowing that a SIREN is nine digits.
function idErrors(ids) {
  if (ids === undefined) return [];
  if (!Array.isArray(ids) || ids.length === 0) return ['legal.ids: non-empty array required when present'];
  const errors = [];
  const seen = new Set();
  ids.forEach((id, i) => {
    if (id === null || typeof id !== 'object' || Array.isArray(id)) {
      errors.push(`legal.ids[${i}]: must be an object`);
      return;
    }
    if (!/^\d{4}$/.test(id.scheme ?? '')) errors.push(`legal.ids[${i}].scheme: 4-digit ICD scheme code required`);
    else if (seen.has(id.scheme)) errors.push(`legal.ids[${i}].scheme: duplicate scheme ${id.scheme}`);
    else seen.add(id.scheme);
    // A number here is the classic hand-rolled-encoder bug: a SIRET emitted as
    // 73282932000074 loses any leading zero and breaks string comparison.
    if (typeof id.value !== 'string' || id.value.trim() === '') errors.push(`legal.ids[${i}].value: non-empty string required`);
  });
  return errors;
}

export function validate(doc) {
  if (doc === null || typeof doc !== 'object' || Array.isArray(doc)) {
    return ['document: must be a JSON object'];
  }
  const errors = [];
  if (!VERSIONS.includes(doc.fpf)) errors.push('fpf: must be "1.1"');
  if (doc.kind !== 'buyer') errors.push('kind: must be "buyer"');

  const legal = doc.legal;
  if (legal === null || typeof legal !== 'object') {
    errors.push('legal: required object');
  } else {
    if (!/^[A-Z]{2}$/.test(legal.country ?? '')) errors.push('legal.country: ISO 3166-1 alpha-2 code required');
    if (typeof legal.name !== 'string' || legal.name.trim() === '') errors.push('legal.name: non-empty string required');
    errors.push(...idErrors(legal.ids));
  }

  const einvoice = doc.einvoice;
  if (einvoice === null || typeof einvoice !== 'object') {
    errors.push('einvoice: required object');
  } else {
    if (!/^\d{4}$/.test(einvoice.eas ?? '')) errors.push('einvoice.eas: 4-digit EAS scheme code required');
    if (typeof einvoice.address !== 'string' || einvoice.address.trim() === '') errors.push('einvoice.address: non-empty string required');
  }

  return errors;
}

File: pascal/src/fpf.pas
unit fpf;

{$MODE OBJFPC}{$H+}

// FPF 1.1 reference implementation — encode/decode of the transport payload.
// Mirrors the Rust reference implementation at ../rust/src/lib.rs.
//
// This unit is portable Object Pascal: everything specific to Free Pascal is
// in fpfjson and fpfbytes, which a Delphi port replaces.

interface

uses
  SysUtils;

type
  // A registration identifier (EN 16931 BT-47) qualified by its ICD scheme
  // code (BT-47-1), drawn from the same registry as einvoice.eas. What a
  // scheme means — 0002 is a French SIREN, 0009 a SIRET — belongs to the
  // country profiles, not here.
  TFpfLegalId = record
    Scheme: string;
    Value: string;
  end;

  TFpfLegalIds = array of TFpfLegalId;

  TFpfLegal = record
    Country: string;
    Name: string;
    Form: string;
    Ids: TFpfLegalIds;
    Vat: string;
    // A dynamic array of length zero IS nil in Pascal, so "present but empty"
    // and "absent" would otherwise be the same thing and the error
    // "legal.ids: non-empty array required when present" could never fire.
    // This flag restores what Option<Vec<LegalId>> carries in Rust. It is only
    // consulted when Ids is empty: a non-empty Ids is present regardless, so
    // an emitter who fills Ids without knowing about the flag never loses it.
    IdsPresent: Boolean;
  end;

  TFpfEinvoice = record
    Eas: string;
    Address: string;
    // Named PlatformName, not Platform: `platform` is a hint directive in
    // Object Pascal. The JSON key is still "platform".
    PlatformName: string;
  end;

  TFpfBilling = record
    Street: string;
    Zip: string;
    City: string;
    Country: string;
  end;

  TFpfContact = record
    Email: string;
    Phone: string;
    // "buyerReference" (EN 16931 BT-10) is the only spelling.
    BuyerReference: string;
  end;

  TFpfDocument = record
    Fpf: string;
    Kind: string;
    Legal: TFpfLegal;
    Einvoice: TFpfEinvoice;
    Billing: TFpfBilling;
    Contact: TFpfContact;
  end;

  TFpfErrors = array of string;

  // The four ways a payload can fail to decode, mirroring Rust's FpfError.
  TFpfErrorKind = (fekUnknownPrefix, fekBase64, fekInflate, fekJson);

  EFpfError = class(Exception)
  public
    Kind: TFpfErrorKind;
    constructor Create(AKind: TFpfErrorKind; const AMessage: string);
  end;

// Serializes and encodes a document into a transport payload. Infallible:
// serializing an already-built record and compressing in memory cannot
// reasonably fail, mirroring the JS and Rust references, which document no
// error path for encoding.
function FpfEncode(const Doc: TFpfDocument; Compress: Boolean = True): string;

// Decodes a transport payload back into a document. Raises EFpfError, whose
// Kind says which of the four ways it failed.
function FpfDecode(const Payload: string): TFpfDocument;

// Structural and semantic validation, mirroring the Rust reference's
// validate(). Unlike the untyped JS reference, the "is this the right shape"
// checks are already guaranteed by holding a TFpfDocument at all: a record
// always has a Legal and an Einvoice, so `legal: required object` has no case
// in which it could fire. A payload missing those keys is refused by
// FpfDecode instead.
function FpfValidate(const Doc: TFpfDocument): TFpfErrors;

implementation

uses
  fpfjson, fpfbytes;

const
  PREFIX_RAW = '1.';
  PREFIX_DEFLATE = '2.';

constructor EFpfError.Create(AKind: TFpfErrorKind; const AMessage: string);
begin
  inherited Create(AMessage);
  Kind := AKind;
end;

function FpfEncode(const Doc: TFpfDocument; Compress: Boolean): string;
var
  Json: RawByteString;
begin
  Json := DocToJson(Doc);
  if Compress then
    Result := PREFIX_DEFLATE + ToBase64Url(RawDeflate(Json))
  else
    Result := PREFIX_RAW + ToBase64Url(Json);
end;

function FpfDecode(const Payload: string): TFpfDocument;
var
  Body: string;
  Bytes: RawByteString;
begin
  if Copy(Payload, 1, Length(PREFIX_DEFLATE)) = PREFIX_DEFLATE then
  begin
    Body := Copy(Payload, Length(PREFIX_DEFLATE) + 1, MaxInt);
    try
      Bytes := FromBase64Url(Body);
    except
      on E: Exception do
        raise EFpfError.Create(fekBase64, 'FPF: base64 decode error: ' + E.Message);
    end;
    try
      Bytes := RawInflate(Bytes);
    except
      on E: Exception do
        raise EFpfError.Create(fekInflate, 'FPF: inflate error: ' + E.Message);
    end;
  end
  else if Copy(Payload, 1, Length(PREFIX_RAW)) = PREFIX_RAW then
  begin
    Body := Copy(Payload, Length(PREFIX_RAW) + 1, MaxInt);
    try
      Bytes := FromBase64Url(Body);
    except
      on E: Exception do
        raise EFpfError.Create(fekBase64, 'FPF: base64 decode error: ' + E.Message);
    end;
  end
  else
    raise EFpfError.Create(fekUnknownPrefix, 'FPF: unknown payload prefix');

  Result := JsonToDoc(Bytes);
end;

function IsAsciiDigits(const S: string; Len: Integer): Boolean;
var
  I: Integer;
begin
  Result := Length(S) = Len;
  if not Result then
    Exit;
  for I := 1 to Length(S) do
    if not (S[I] in ['0'..'9']) then
      Exit(False);
end;

function IsCountryCode(const S: string): Boolean;
var
  I: Integer;
begin
  Result := Length(S) = 2;
  if not Result then
    Exit;
  for I := 1 to Length(S) do
    if not (S[I] in ['A'..'Z']) then
      Exit(False);
end;

procedure Add(var Errors: TFpfErrors; const Message: string);
begin
  SetLength(Errors, Length(Errors) + 1);
  Errors[High(Errors)] := Message;
end;

function FpfValidate(const Doc: TFpfDocument): TFpfErrors;
var
  I, J: Integer;
  Duplicate: Boolean;
begin
  Result := nil;

  if Doc.Fpf <> '1.1' then
    Add(Result, 'fpf: must be "1.1"');
  if Doc.Kind <> 'buyer' then
    Add(Result, 'kind: must be "buyer"');

  if not IsCountryCode(Doc.Legal.Country) then
    Add(Result, 'legal.country: ISO 3166-1 alpha-2 code required');
  if Trim(Doc.Legal.Name) = '' then
    Add(Result, 'legal.name: non-empty string required');

  if (Length(Doc.Legal.Ids) > 0) or Doc.Legal.IdsPresent then
  begin
    if Length(Doc.Legal.Ids) = 0 then
      Add(Result, 'legal.ids: non-empty array required when present');
    for I := 0 to High(Doc.Legal.Ids) do
    begin
      if not IsAsciiDigits(Doc.Legal.Ids[I].Scheme, 4) then
        Add(Result, Format('legal.ids[%d].scheme: 4-digit ICD scheme code required', [I]))
      else
      begin
        Duplicate := False;
        for J := 0 to I - 1 do
          if Doc.Legal.Ids[J].Scheme = Doc.Legal.Ids[I].Scheme then
            Duplicate := True;
        if Duplicate then
          Add(Result, Format('legal.ids[%d].scheme: duplicate scheme %s', [I, Doc.Legal.Ids[I].Scheme]));
      end;
      // A number here is the classic hand-rolled-encoder bug: a SIRET emitted
      // as 73282932000074 loses any leading zero and breaks string comparison.
      if Trim(Doc.Legal.Ids[I].Value) = '' then
        Add(Result, Format('legal.ids[%d].value: non-empty string required', [I]));
    end;
  end;

  if not IsAsciiDigits(Doc.Einvoice.Eas, 4) then
    Add(Result, 'einvoice.eas: 4-digit EAS scheme code required');
  if Trim(Doc.Einvoice.Address) = '' then
    Add(Result, 'einvoice.address: non-empty string required');

end;

end.
